File: .claude/hooks/protect-lockfiles.mjs
#!/usr/bin/env node
// @hookstack pre-edit-protect-lockfiles
// Bloque les éditions directes de fichiers de lock (PreToolUse Write|Edit)
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';

const LOCKFILES = [
  /pnpm-lock\.yaml$/,
  /package-lock\.json$/,
  /yarn\.lock$/,
  /Gemfile\.lock$/,
  /poetry\.lock$/,
  /Pipfile\.lock$/,
  /composer\.lock$/,
  /Cargo\.lock$/,
];

export function run(input) {
  const filePath = input.tool_input?.file_path ?? '';
  if (!filePath) return null;
  const match = LOCKFILES.find((p) => p.test(filePath));
  return match
    ? {
        decision: 'block',
        reason: `Lock file ${filePath} must not be edited directly. Run the package manager instead (pnpm install, npm install, cargo build…).`,
      }
    : null;
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/protect-lockfiles.py
#!/usr/bin/env python3
# @hookstack pre-edit-protect-lockfiles
"""Blocks direct edits of lock files (PreToolUse Write|Edit)."""
import json
import re
import sys

LOCKFILES = [
    re.compile(r"pnpm-lock\.yaml$"),
    re.compile(r"package-lock\.json$"),
    re.compile(r"yarn\.lock$"),
    re.compile(r"Gemfile\.lock$"),
    re.compile(r"poetry\.lock$"),
    re.compile(r"Pipfile\.lock$"),
    re.compile(r"composer\.lock$"),
    re.compile(r"Cargo\.lock$"),
]


def run(input_data):
    file_path = (input_data.get("tool_input") or {}).get("file_path") or ""
    if not file_path:
        return None
    for pattern in LOCKFILES:
        if pattern.search(file_path):
            return {
                "decision": "block",
                "reason": (
                    f"Lock file {file_path} must not be edited directly. Run the package manager "
                    "instead (pnpm install, npm install, cargo build…)."
                ),
            }
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
